src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,317 @@
+ */
+class TokenCacheAccessor {
+    private static final String TAG = TokenCacheAccessor.class.getSimpleName();
+    
+    private final ITokenCacheStore mTokenCacheStore;
+    private final String mAuthority;
+    
+    TokenCacheAccessor(final ITokenCacheStore tokenCacheStore, final String authority) {
+        if (tokenCacheStore == null) {
+            throw new IllegalArgumentException("tokenCacheStore");
+        }
+        
+        if (StringExtensions.IsNullOrBlank(authority)) {
+            throw new IllegalArgumentException("authority");
+        }
+        
+        mTokenCacheStore = tokenCacheStore;
+        mAuthority = authority;
+    }
+    
+    /**
+     * @return Access token from cache. Could be null if AT does not exist or expired. 
+     * This will be a strict match with the user passed in, could be unique userid, 
+     * displayable id, or null user. 
+     * @throws AuthenticationException 
+     */
+    TokenCacheItem getATFromCache(final String resource, final String clientId, final String user) 
+            throws AuthenticationException {
+        final TokenCacheItem accessTokenItem = getRegularRefreshTokenCacheItem(resource, clientId, user);
+        if (accessTokenItem == null) {
+            Logger.v(TAG, "No access token exists.");
+            return null;
+        }
+        
+        if (!StringExtensions.IsNullOrBlank(accessTokenItem.getAccessToken())) {
+            if (accessTokenItem.isTokenExpired(accessTokenItem.getExpiresOn())) {
+                Logger.v(TAG, "Access token exists, but already expired.");
+                return null;
+            }
+            
+            // To support backward-compatibility, for old token entry, user stored in 
+            // token cache item could be different from the one in cachekey. 
+            if (isUserMisMatch(user, accessTokenItem)) {
+                throw new AuthenticationException(ADALError.AUTH_FAILED_USER_MISMATCH);
+            }
+        }
+        
+        return accessTokenItem;
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for regular token cache entry.  
+     */
+    TokenCacheItem getRegularRefreshTokenCacheItem(final String resource, final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForRTEntry(mAuthority, resource, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for MRRT token cache entry.  
+     */
+    TokenCacheItem getMRRTItem(final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForMRRT(mAuthority, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for FRT token cache entry.  
+     */
+    TokenCacheItem getFRTItem(final String familyClientId, final String user) {
+        if (StringExtensions.IsNullOrBlank(user)) {
```

Comment:
Nope. FamilyRT is under such a loose cachekey criteria an is the best effort we can do.  We want to be conservative for FRT lookup, use it only when either uniqueId or displayable id is provided. 

---
In reply to: [63256037](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63256037) [](ancestors = 63256037)